src/backend/d3d12/BufferD3D12.cpp

```diff
@@ -99,16 +99,43 @@ namespace d3d12 {
         return resource->GetGPUVirtualAddress();
     }
 
+    void Buffer::OnMapReadCommandSerialFinished(uint32_t mapSerial, uint32_t count) {
+        D3D12_RANGE readRange = { 0, count };
+        char* data;
+        readbackResource->Map(0, &readRange, reinterpret_cast<void**>(&data));
+        CallMapReadCallback(mapSerial, NXT_BUFFER_MAP_READ_STATUS_SUCCESS, data);
+    }
+
     void Buffer::SetSubDataImpl(uint32_t start, uint32_t count, const uint32_t* data) {
         device->GetResourceUploader()->BufferSubData(resource, start * sizeof(uint32_t), count * sizeof(uint32_t), data);
     }
 
     void Buffer::MapReadAsyncImpl(uint32_t serial, uint32_t start, uint32_t count) {
-        // TODO(cwallez@chromium.org): Implement Map Read for the null backend
+        D3D12_RESOURCE_DESC resourceDescriptor;
+        resourceDescriptor.Dimension = D3D12_RESOURCE_DIMENSION_BUFFER;
+        resourceDescriptor.Alignment = 0;
+        resourceDescriptor.Width = count;
+        resourceDescriptor.Height = 1;
+        resourceDescriptor.DepthOrArraySize = 1;
+        resourceDescriptor.MipLevels = 1;
+        resourceDescriptor.Format = DXGI_FORMAT_UNKNOWN;
+        resourceDescriptor.SampleDesc.Count = 1;
+        resourceDescriptor.SampleDesc.Quality = 0;
+        resourceDescriptor.Layout = D3D12_TEXTURE_LAYOUT_ROW_MAJOR;
+        resourceDescriptor.Flags = D3D12_RESOURCE_FLAG_NONE;
+
+        readbackResource = device->GetResourceAllocator()->Allocate(D3D12_HEAP_TYPE_READBACK, resourceDescriptor, D3D12_RESOURCE_STATE_COPY_DEST);
```

Comment:
A buffer that has a MapRead usage can be created in the READBACK heap immediately, so there is no need to create a readbackResource nor do a copy.